nav2_costmap_2d/test/integration/test_collision_checker.cpp

```diff
@@ -0,0 +1,328 @@
+      "GetRobotPose", std::bind(&TestCollisionChecker::get_robot_pose_callback, this, _1, _2, _3));
+
+    layers_ = new nav2_costmap_2d::LayeredCostmap("frame", false, false);
+    // Add Static Layer
+    addStaticLayer(*layers_, *tf_buffer_, shared_from_this());
+    // Add Inflation Layer
+    addInflationLayer(*layers_, *tf_buffer_, shared_from_this());
+
+    return nav2_util::CallbackReturn::SUCCESS;
+  }
+
+  nav2_util::CallbackReturn
+  on_activate(const rclcpp_lifecycle::State & /*state*/)
+  {
+    RCLCPP_INFO(get_logger(), "Activating");
+    return nav2_util::CallbackReturn::SUCCESS;
+  }
+
+  nav2_util::CallbackReturn
+  on_deactivate(const rclcpp_lifecycle::State & /*state*/)
+  {
+    RCLCPP_INFO(get_logger(), "Deactivating");
+    return nav2_util::CallbackReturn::SUCCESS;
+  }
+
+  nav2_util::CallbackReturn
+  on_cleanup(const rclcpp_lifecycle::State & /*state*/)
+  {
+    RCLCPP_INFO(get_logger(), "Cleaning Up");
+    delete layers_;
+    layers_ = nullptr;
+
+    tf_buffer_.reset();
+    tf_listener_.reset();
+
+    footprint_sub_.reset();
+    costmap_sub_.reset();
+
+    return nav2_util::CallbackReturn::SUCCESS;
+  }
+
+  ~TestCollisionChecker() {}
+
+  bool testPose(double x, double y, double theta)
+  {
+    geometry_msgs::msg::Pose2D pose;
+    pose.x = x;
+    pose.y = y;
+    pose.theta = theta;
+
+    setPose(x, y, theta);
+    publishFootprint();
+    publishCostmap();
+    return collision_checker_->isCollisionFree(pose);
+  }
+
+  void setFootprint(double footprint_padding, double robot_radius)
+  {
+    std::vector<geometry_msgs::msg::Point> new_footprint;
+    new_footprint = nav2_costmap_2d::makeFootprintFromRadius(robot_radius);
+    nav2_costmap_2d::padFootprint(new_footprint, footprint_padding);
+    footprint_ = new_footprint;
+    layers_->setFootprint(footprint_);
+  }
+
+protected:
+  void setPose(double x, double y, double theta)
+  {
+    x_ = x;
+    y_ = y;
+    yaw_ = theta;
+
+    current_pose_.pose.position.x = x_;
+    current_pose_.pose.position.y = y_;
+    current_pose_.pose.position.z = 0;
+    tf2::Quaternion q;
+    q.setRPY(0, 0, yaw_);
+    current_pose_.pose.orientation = tf2::toMsg(q);
+  }
+
+  void publishFootprint()
+  {
+    geometry_msgs::msg::PolygonStamped oriented_footprint;
+    oriented_footprint.header.frame_id = global_frame_;
+    oriented_footprint.header.stamp = now();
+    nav2_costmap_2d::transformFootprint(x_, y_, yaw_, footprint_, oriented_footprint);
+    footprint_sub_->setFootprint(
+      std::make_shared<geometry_msgs::msg::PolygonStamped>(oriented_footprint));
+  }
+
+  void publishCostmap()
+  {
+    layers_->updateMap(x_, y_, yaw_);
+    costmap_sub_->setCostmap(
+      std::make_shared<nav2_msgs::msg::Costmap>(toCostmapMsg(layers_->getCostmap())));
+  }
+
+  nav2_msgs::msg::Costmap
+  toCostmapMsg(nav2_costmap_2d::Costmap2D * costmap)
+  {
+    double resolution = costmap->getResolution();
```

Comment:
A personal, stylistic thing, but I'd organize this code to get the various values first, and then populate the structure.  For example, 

```C++
    double resolution = costmap->getResolution();

    double wx, wy;
    costmap->mapToWorld(0, 0, wx, wy);

    unsigned char * data = costmap->getCharMap();

    nav2_msgs::msg::Costmap costmap_msg;
    costmap_msg.header.frame_id = global_frame_;
    costmap_msg.header.stamp = now();
    costmap_msg.metadata.layer = "master";
    costmap_msg.metadata.resolution = resolution;
    costmap_msg.metadata.size_x = costmap->getSizeInCellsX();
    costmap_msg.metadata.size_y = costmap->getSizeInCellsY();
    costmap_msg.metadata.origin.position.x = wx - resolution / 2;
    costmap_msg.metadata.origin.position.y = wy - resolution / 2;
    costmap_msg.metadata.origin.position.z = 0.0;
    costmap_msg.metadata.origin.orientation.w = 1.0;
    costmap_msg.data.resize(costmap_msg.metadata.size_x * costmap_msg.metadata.size_y);

    for (unsigned int i = 0; i < costmap_msg.data.size(); i++) {
      costmap_msg.data[i] = data[i];
    }

     return costmap_msg;
```

I think the result is a bit easier to read and to see all of the structure fields.

Reply:
Sure, I like that.